extensions/src/main/scala/ai/h2o/sparkling/extensions/rest/api/ChunkServlet.scala

```diff
@@ -17,113 +17,131 @@
 
 package ai.h2o.sparkling.extensions.rest.api
 
-import java.nio.{ByteBuffer, ByteOrder}
-
 import ai.h2o.sparkling.utils.ScalaUtils._
 import ai.h2o.sparkling.utils.Base64Encoding
-import ai.h2o.sparkling.extensions.serde.{SerializationUtils, ChunkAutoBufferWriter}
+import ai.h2o.sparkling.extensions.serde.{ChunkAutoBufferReader, ChunkAutoBufferWriter, ChunkSerdeConstants}
 import javax.servlet.http.{HttpServlet, HttpServletRequest, HttpServletResponse}
 import water.DKV
 import water.fvec.Frame
 import water.server.ServletUtils
 
 /**
-  * This servlet class handles GET requests for the path /3/Chunk
-  * It requires 4 get parameters
-  * - frame_name - a unique string identifier of H2O Frame
-  * - chunk_id - a unique identifier of the chunk within the H2O Frame
-  * - expected_type - byte array encoded in Base64 encoding. The types corresponds to the `selected_columns` parameter
-  * - selected_columns - selected columns indices encoded int Base64 encoding.
-  * The result is represented as a stream of binary data. Data are encoded to AutoBuffer row by row.
-  * The data stream starts with the integer representing the number of rows.
+  * This servlet class handles GET and PUT requests for the path /3/Chunk
   */
 final class ChunkServlet extends HttpServlet {
 
-  private case class RequestParameters(
+  private case class GetRequestParameters(
       frameName: String,
       chunkId: Int,
       expectedTypes: Array[Byte],
-      selectedColumnIndices: Array[Int])
+      selectedColumnIndices: Array[Int]) {
 
-  private def getParameterAsString(request: HttpServletRequest, parameterName: String): String = {
-    val result = request.getParameter(parameterName)
-    if (result == null) {
-      throw new RuntimeException(s"Cannot find value for the parameter '$parameterName'")
+    def validate(): Unit = {
+      val frame = DKV.getGet[Frame](this.frameName)
+      if (frame == null) throw new RuntimeException(s"A frame with name '$frameName")
```

Comment:
Also just an idea: Would it be possible to return the exception message to the client in case something went wrong? We could store the exception into the `response` and just create a new schema type like H2OExceptionV3.

It would be nice if all errors on the backend via rest api are propagated to the client as well

Reply:
IMHO, the errors will be propagated to the client anyway: https://github.com/h2oai/sparkling-water/pull/1783/files/8ae85fe91cd91424eb3dd89c03af43f5b64dd651#diff-77ccca4e517cc75d148a17f89f920188R124

I hope I don't miss anything, need to test that.

Reply:
Perfect!